### PR TITLE
Add cpu target percentile parameter to VPA options in Shoot spec

### PR DESCRIPTION
**Pull request by voelzmo**
**How to categorize this PR?**

/area auto-scaling
/kind enhancement

**What this PR does / why we need it**:
This PR exposes the `targetCPUPercentile` parameter of the VPA recommender in the Shoot spec. With this parameter, Shoot owners or gardener operators can fine-tune the recommender behavior regarding recommendations for CPU.

The default is 0.9, which means the recommender will issue a recommendation that is fulfilling 90% of the observed CPU measurements. As CPU is a compressible resource, Shoot owner may want to set this to a smaller value, depending on the headroom on their nodes and the requirements for their workloads.

**Release note**:

```feature user
In order to fine-tune VPA CPU recommendations, the CPU target percentile parameter for `vpa-recommender` can now be configured in the `Shoot` specification via the `.spec.kubernetes.verticalPodAutoscaler.targetCPUPercentile` field.
```

### CODE REVIEW

File: pkg/apis/core/types_shoot.go [hunk 1]
```diff
@@ -536,6 +536,11 @@ type VerticalPodAutoscaler struct {
 	UpdaterInterval *metav1.Duration
 	// RecommenderInterval is the interval how often metrics should be fetched (default: 1m0s).
 	RecommenderInterval *metav1.Duration
+	// TargetCPUPercentile is the usage percentile that will be used as a base for CPU target recommendation.
+	// Doesn't affect CPU lower bound, CPU upper bound nor memory recommendations.
+	// (default: 0.9)
+	// +optional
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]:
```suggestion
```

not needed in internal API

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 2]
```diff
@@ -662,6 +662,11 @@ type VerticalPodAutoscaler struct {
 	// RecommenderInterval is the interval how often metrics should be fetched (default: 1m0s).
 	// +optional
 	RecommenderInterval *metav1.Duration `json:"recommenderInterval,omitempty" protobuf:"bytes,8,opt,name=recommenderInterval"`
+	// TargetCPUPercentile is the usage percentile that will be used as a base for CPU target recommendation.
+	// Doesn't affect CPU lower bound, CPU upper bound nor memory recommendations.
+	// (default: 0.9)
+	// +optional
+	TargetCPUPercentile *float64 `json:"targetCPUPercentile,omitempty" protobuf:"fixed64,9,opt,name=targetCPUPercentile"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go:671]: Please enhance https://github.com/gardener/gardener/blob/76704c377f34cdbdf1b0d3986b243c8b67c66909/pkg/apis/core/v1beta1/defaults_shoot.go#L255-L285 and https://github.com/gardener/gardener/blob/54b8f179143347e5143c38ae761a90bb63f9ecba/example/90-shoot.yaml#L298-L306 and https://github.com/gardener/gardener/blob/master/docs/usage/shoot_autoscaling.md#vertical-pod-auto-scaling

Comment (voelzmo) [reply pkg/apis/core/v1beta1/types_shoot.go:671]: done.

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 3]
```diff
@@ -675,6 +680,8 @@ const (
 	DefaultEvictionTolerance = 0.5
 	// DefaultRecommendationMarginFraction is the default value for the RecommendationMarginFraction field in the VPA configuration.
 	DefaultRecommendationMarginFraction = 0.15
+	// DefaultTargetCPUPercentile is the default calue for the TargetCPUPercentile field in the VPA configuration
```
Comment (ary1992) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// DefaultTargetCPUPercentile is the default value for the TargetCPUPercentile field in the VPA configuration
```

**Review by ary1992 — CHANGES_REQUESTED** [review]

File: pkg/apis/core/v1beta1/types_shoot.go (hunk 3 above)
Comment (voelzmo) [reply pkg/apis/core/v1beta1/types_shoot.go]: done, thanks!

**Comment by voelzmo** [comment]
/retest
/hold cancel
Tests have been successful: when modifying the parameter, we see that his has a direct impact on the cpu recommendations. Lower values lead to lower recommendations overall.

**Comment by voelzmo** [comment]
> PR looks good, but adaptation in https://github.com/gardener/gardener/blob/master/docs/usage/shoot_autoscaling.md#vertical-pod-auto-scaling is still missing.

Ah, sorry, thanks for catching that I missed that.
On a related note: I just find myself copy-pasting the same things that I already wrote in the API docs for those properties. Should we maybe just link to the API docs instead of replicating that information here?

**Comment by rfranzke** [comment]
> Should we maybe just link to the API docs instead of replicating that information here?

We could. OTOH, the API docs typically do not contain too much information. Maybe this works in case of the VPA fields - if so, feel free to just link the generated docs (maybe in a separate PR).

/lgtm
/approve

**Pull request merged**